src/esphome/sensor/ppd42x.cpp

```diff
@@ -0,0 +1,122 @@
+
+void PPD42XComponent::loop() {
+  const uint32_t now = millis();
+  uint32_t duration_pl_02_5 = pulseIn(this->pl_02_5_sensor_->pl_pin_->get_pin(),
+                                      uint8_t(!this->pl_02_5_sensor_->pl_pin_->is_inverted()),
+              this->timeout_us_);
+  uint32_t duration_pl_10_0 = pulseIn(this->pl_10_0_sensor_->pl_pin_->get_pin(),
+                                      uint8_t(!this->pl_10_0_sensor_->pl_pin_->is_inverted()),
+              this->timeout_us_);
+  this->lowpulseoccupancy_02_5_ = this->lowpulseoccupancy_02_5_ + duration_pl_02_5;
+  this->lowpulseoccupancy_10_0_ = this->lowpulseoccupancy_10_0_ + duration_pl_10_0;
+
+  if (now - this->starttime_ > this->timeout_us_) {
+    // last transmission too long ago
+    this->starttime_ = now;
+    parse_data_();
+  }
+}
+float PPD42XComponent::get_setup_priority() const { return setup_priority::HARDWARE_LATE; }
+
+void PPD42XComponent::parse_data_() {
+  switch (this->ctype_) {
+    case PPD42X_TYPE: {
+      float pl_02_5_concentration = us_to_pl(this->lowpulseoccupancy_02_5_, this->timeout_us_);
+      float pl_10_0_concentration = us_to_pl(this->lowpulseoccupancy_10_0_, this->timeout_us_);
+      ESP_LOGD(TAG, "Got PM2.5 Concentration %f pcs/L, PM10.0 Concentration: %f pcs/L", pl_02_5_concentration,
+               pl_10_0_concentration);
+      if (this->pl_02_5_sensor_ != nullptr)
+        this->pl_02_5_sensor_->publish_state(pl_02_5_concentration);
+      if (this->pl_10_0_sensor_ != nullptr)
+        this->pl_10_0_sensor_->publish_state(pl_10_0_concentration);
+      break;
+    }
+    case PPD42X_TYPE_NS: {
+      float pl_02_5_concentration = us_to_pl(this->lowpulseoccupancy_02_5_, this->timeout_us_);
+      float pl_10_0_concentration = us_to_pl(this->lowpulseoccupancy_10_0_, this->timeout_us_);
+      ESP_LOGD(TAG, "Got PM2.5 Concentration %f pcs/L, PM10.0 Concentration: %f pcs/L", pl_02_5_concentration,
+               pl_10_0_concentration);
+      if (this->pl_02_5_sensor_ != nullptr)
+        this->pl_02_5_sensor_->publish_state(pl_02_5_concentration);
+      if (this->pl_10_0_sensor_ != nullptr)
+        this->pl_10_0_sensor_->publish_state(pl_10_0_concentration);
+      break;
+    }
+    case PPD42X_TYPE_NJ: {
+      float pl_02_5_concentration = us_to_pl(this->lowpulseoccupancy_02_5_, this->timeout_us_);
+      float pl_10_0_concentration = us_to_pl(this->lowpulseoccupancy_10_0_, this->timeout_us_);
+      ESP_LOGD(TAG, "Got PM2.5 Concentration %f pcs/L, PM10.0 Concentration: %f pcs/L", pl_02_5_concentration,
+               pl_10_0_concentration);
+      if (this->pl_02_5_sensor_ != nullptr)
+        this->pl_02_5_sensor_->publish_state(pl_02_5_concentration);
+      if (this->pl_10_0_sensor_ != nullptr)
+        this->pl_10_0_sensor_->publish_state(pl_10_0_concentration);
+      break;
+    }
+  }
+}
+float PPD42XComponent::us_to_pl(uint32_t sample_length, uint32_t time_pm) {
+  float ratio = time_pm / (sample_length * 10.0f);
```

Comment:
Why does this depend on the timeout value?

Reply:
this is the recommended max value to listen to !

Reply:
I mean why does this _calculation_ depend on the timeout? The sensor doesn't know which timeout we chose, so how can this calculation depend on the timeout value `this->timeout_us_` we chose?

Reply:
nop, the sensor send continuously data, and we cannot stay stuck too long to lisning the end of the signal, so we timout...